quic/connection.nim

```diff
@@ -14,22 +12,22 @@ type
     udp: DatagramTransport
     quic: QuicConnection
     loop: Future[void]
-    onClose: proc() {.gcsafe.}
+    onClose: ?proc() {.gcsafe.}
     closed: AsyncEvent
   IncomingConnection = ref object of Connection
   OutgoingConnection = ref object of Connection
 
 proc ids*(connection: Connection): seq[ConnectionId] =
   connection.quic.ids
 
-proc `onNewId=`*(connection: Connection, callback: proc(id: ConnectionId)) =
+proc `onNewId=`*(connection: Connection, callback: IdCallback) =
   connection.quic.onNewId = callback
 
-proc `onRemoveId=`*(connection: Connection, callback: proc(id: ConnectionId)) =
+proc `onRemoveId=`*(connection: Connection, callback: IdCallback) =
   connection.quic.onRemoveId = callback
 
 proc `onClose=`*(connection: Connection, callback: proc() {.gcsafe.}) =
-  connection.onClose = callback
+  connection.onClose = some callback
```

Comment:
I'm not sure what the improvement is here - ie `callback` can still be nil, meaning it's simply two opportunities for failure instead of one

Reply:
ie until there's `not nil` support in the language, this is just a more expensive way of saying the same thing, and ultimately, having the same errors - callbacks, or any `proc`/`ref` types can always be `nil` - this is already idiomatic in the language itself, just as strongly as `Option` can be unset

Reply:
In some places in the code, references or pointers were used to to signal the presence of a value. In this case, the onClose callback can either be absent or present. It is however easy to forget to check for a nil value. By wrapping it in an Option, the compiler forces me to do the check.

Now, the callback inside the Option can still be nil, but that would require an explicit assignment of nil. If we want to guard against that, we could change this line to:

```nim
connection.onClose = callback.option
```

Reply:
I give Nim some credit here, as `option` is a bit more sneaky than that.  From `Nim/lib/pure/options.nim`:

```nim
  when T is SomePointer:
    assert(not val.isNil)
    result.val = val
  else:
    result.has = true
    result.val = val
```

So it's not possible to have `some nil`.  You can ignore the fact `assert` may be compiled out: If `nil` is stored, the option will satisfy `isNone` not `isSome`.

Sometimes there's an advantage to this:

- The optional type is a kind of documentation that it's fine for _this_ `proc`/`ref` to be unset, unlike others in the code.
- The type system forces explicit checking for `None` before each use (unlike `nil`).
- The default value is `None` (i.e. not `nil`).
- Even if `some nil` -> `isSome()` could be true, it would be less likely to occur by mistake compared with plain `nil` just because the latter is the default, and the former would have to be explicitly stored, which is unlikely in uses like `c.onClose(someNamedProc)`.

(Before looking at `options.nim` I was inclined to agree that the option is expensive overkill even with documentary benefits, I'd just put a `nil` check where `connection.onClose` is called instead, if it's meant to be optional.)

Reply:
good points, both :+1:

Reply:
the remaining "problem" being that any non-option use of `ref`/`proc` remains ambiguous and must be checked for `nil` unless mentioned otherwise, so there's still the cognitive overhead of keeping track of these nuances - hm, maybe using `Option` should become a style recommendation for _all_ cases that admit `nil`

Reply:
> maybe using Option should become a style recommendation for all cases that admit nil 

Not a bad idea at all. @markspanbroek and @jlokier have solid arguments to make this into a default across the board.